test/integration/scheduler_test.go

```diff
@@ -103,50 +113,171 @@ func podScheduled(c *client.Client, podNamespace, podName string) wait.Condition
+	return wait.Poll(time.Second, time.Second*20, func() (bool, error) {
+		if n, _, err := s.GetByKey(key); err == nil && passFunc(n) {
+			return true, nil
+		}
+		return false, nil
+	})
+}
+
+func DoTestUnschedulableNodes(t *testing.T, restClient *client.Client, nodeStore cache.Store) {
+	goodCondition := api.NodeCondition{
+		Type:              api.NodeReady,
+		Status:            api.ConditionTrue,
+		Reason:            fmt.Sprintf("schedulable condition"),
+		LastHeartbeatTime: util.Time{time.Now()},
 	}
-	if _, err := client.Nodes().Create(node); err != nil {
-		t.Fatalf("Failed to create node: %v", err)
+	badCondition := api.NodeCondition{
+		Type:              api.NodeReady,
+		Status:            api.ConditionUnknown,
+		Reason:            fmt.Sprintf("unschedulable condition"),
+		LastHeartbeatTime: util.Time{time.Now()},
 	}
-
-	pod := &api.Pod{
-		ObjectMeta: api.ObjectMeta{Name: "my-pod"},
-		Spec: api.PodSpec{
-			Containers: []api.Container{{Name: "container", Image: "kubernetes/pause:go"}},
+	// Create a new schedulable node, since we're first going to apply
+	// the unschedulable condition and verify that pods aren't scheduled.
+	node := &api.Node{
+		ObjectMeta: api.ObjectMeta{Name: "node"},
+		Spec:       api.NodeSpec{Unschedulable: false},
+		Status: api.NodeStatus{
+			Conditions: []api.NodeCondition{goodCondition},
 		},
 	}
-	myPod, err := client.Pods(api.NamespaceDefault).Create(pod)
+	nodeKey, err := cache.MetaNamespaceKeyFunc(node)
 	if err != nil {
-		t.Fatalf("Failed to create pod: %v", err)
-	}
-	// There are no schedulable nodes - the pod shouldn't be scheduled.
-	err = wait.Poll(time.Second, time.Second*10, podScheduled(client, myPod.Namespace, myPod.Name))
-	if err == nil {
-		t.Errorf("Pod scheduled successfully on unschedulable nodes")
-	}
-	if err != wait.ErrWaitTimeout {
-		t.Errorf("Failed while waiting for scheduled pod: %v", err)
+		t.Fatalf("Couldn't retrieve key for node %v", node.Name)
 	}
 
-	// Make the node schedulable and wait until the pod is scheduled.
-	newNode, err := client.Nodes().Get(node.Name)
-	if err != nil {
-		t.Fatalf("Failed to get node: %v", err)
-	}
-	newNode.Spec.Unschedulable = false
-	if _, err = client.Nodes().Update(newNode); err != nil {
-		t.Fatalf("Failed to update node: %v", err)
-	}
-	err = wait.Poll(time.Second, time.Second*10, podScheduled(client, myPod.Namespace, myPod.Name))
-	if err != nil {
-		t.Errorf("Failed to schedule a pod: %v", err)
+	// The test does the following for each nodeStateManager in this list:
+	//	1. Create a new node
+	//	2. Apply the unschedulable function
+	//	3. Create a new pod
+	//  4. Check that the pod doesn't get assigned to the node
+	//  5. Apply the schedulable function
+	//  6. Check that the pod *does* get assigned to the node
+	//  7. Delete the pod and node.
+
+	nodeModifications := []nodeStateManager{
+		// Test node.Spec.Unschedulable=true/false
+		{
+			unschedulable: func(t *testing.T, n *api.Node, s cache.Store, c *client.Client) {
```

Comment:
Suggest make these functions just modify a node-- the rest is duplicated boilerplate that can be done in the test loop.